### PR TITLE
Design document

### PR DESCRIPTION
**Pull request by tolikzinovyev**
## Content

The first design document for our implementation. Superceeds #29.

### CODE REVIEW

**Comment by tolikzinovyev** [comment]
@cardano-scaling/team-alba please give your high level feedback since I haven't polished all details. Thanks.

File: design/design.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+### Overview of ALBA
+ALBA is a cryptographic primitive that allows one to prove knowledge of
+some number of data elements of some kind.
+For example, one can prove knowledge of $n$ cryptographic signatures signing
+the same message, effectively constructing a multisignature scheme.
+ALBA, however, requires that there is a gap between what the prover knows and
+what the verifier is convinced the prover knows.
+Specifically, given parameters $n_f$ and $n_p$ such that $n_f < n_p$,
+the prover who possesses $n_p$ element of interest is able to convince
```
Comment (djetchev) [inline design/design.md:26]: $n_p$ elements

File: design/design.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+the verifier that he knows $> n_f$ elements of interest.
+Hence, the lower bound argument is *approximate*.
+The larger the ratio $n_p / n_f$ is, the smaller the ALBA certificate is.
+
+ALBA can also work in the weighted regime where data elements can have
+an integer weight.
```
Comment (djetchev) [inline design/design.md:32]: To relate to specific applications, maybe say that this is particularly useful for PoS protocols and stake-based signatures.

File: design/design.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+Compared to SNARKs that can solve the same problem, ALBA constructions offer
+much faster proving times: in the order of a second.
+Additionally, the decentralized Telescope ALBA construction gives rise to
+the only known multisignature scheme with a configurable tradeoff between
+proof size and (sublinear) communication complexity.
+The only other known way to achieve sublinear communication complexity is
+through a lottery as used, for example, by Algorand's sortition or the Mithril
+protocol (todo: add link).
+The lottery has slightly better communication complexity but a significantly
```
Comment (djetchev) [inline design/design.md:58]: I wonder if it makes sense to add some comparisons here - to show the overhead on proving times.

**Comment by tolikzinovyev** [comment]
@djetchev I will address your comments, but do you have feedback about the overall structure first?

File: design/design.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+parameters in their application.
+The ALBA functions such as `prove` and `verify` will thus receive as input
+the algorithm specific parameters.
+
+The rationale is that we want to be able to evolve parameter derivation
+(e.g., to optimize parameters for performance or make the derivation logic
+more rigorous) while maintaining backward compatibility with previous versions
+of the library.
+
+## Roadmap
```
Comment (curiecrypt) [inline design/design.md:393]: I think it's a good practice to group the goals of the roadmap.
For example:
- v0: Aim to finish the schemes we mentioned in the ALBA algorithms section,
- v1: generic hash function, elements, benchmarks,
- v2: ...
- ...
Moreover, we could name those as some checkpoints with IDs or some names (like the eras of Cardano). And, state the objectives that we want to achieve for each era. Even, we can add an actual image of the roadmap to visualize it better.

**Review by curiecrypt — COMMENTED** [review]
Looks much better in this version! I will go for another round after the **todo**s are completed.
An aesthetic suggestion would be to use **bold** and _italic_ style for some important points, i.e., 1) Additionally, ALBA supports a _decentralized_ setting, 2) This cryptographic primitive has a number of applications, in the **Cardano** space
and outside.
Also, using bullets where needed might increase the readability and the reader could focus more easily.

File: design/design.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+[@tolikzinovyev](https://github.com/tolikzinovyev) November 2024
+
+## Introduction
+
+This is the first design document for our Rust implementation of
+Approximate Lower Bound Arguments (ALBA) intended to guide the development
+of the library.
+We aim to create a high quality reference implementation of ALBA sufficient
+for accurately prototyping larger protocols utilizing this primitive.
+Applications that could benefit, at least in the Cardano universe, include
```
Comment (rrtoledo) [inline design/design.md:12]: Add also general examples for users outside of Cardano

File: design/design.md [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+Mithril, Peras, Leios and cross-chain bridges.
+These use cases will generally guide the development of our library, though
+our aim is to also allow the open-source community to study and experiment
+with ALBA.
+
+### Overview of ALBA
+ALBA is a cryptographic primitive that allows one to prove knowledge of
+some number of data elements of some kind.
```
Comment (rrtoledo) [inline design/design.md:20]: This is way to vague, ALBA is a proof of lower bound, so just write it as so.

File: design/design.md [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+For example, one can prove knowledge of $n$ cryptographic signatures signing
+the same message, effectively constructing a multisignature scheme.
+ALBA, however, requires that there is a gap between what the prover knows and
```
Comment (rrtoledo) [inline design/design.md:23]: You wouldn't have to write this if you say that ALBA is a lower bound argument

File: design/design.md [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+what the verifier is convinced the prover knows.
+Specifically, given parameters $n_f$ and $n_p$ such that $n_f < n_p$,
+the prover who possesses $n_p$ element of interest is able to convince
+the verifier that he knows $> n_f$ elements of interest.
```
Comment (rrtoledo) [inline design/design.md:27]: Write in english instead of `>`

File: design/design.md [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+Hence, the lower bound argument is *approximate*.
```
Comment (rrtoledo) [inline design/design.md:28]: Can you clarify why it is "approximate"? It is explicit.

File: design/design.md [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+The larger the ratio $n_p / n_f$ is, the smaller the ALBA certificate is.
+
+ALBA can also work in the weighted regime where data elements can have
+an integer weight.
+Then the prover who possesses elements of total weight $\ge n_p$ is able to
+convince the verifier that he knows elements of total weight $> n_f$.
+Such a gap between $n_p$ and $n_f$ exists naturally in secure distributed
```
Comment (rrtoledo) [inline design/design.md:35]: I am not sure it is relevant to talk about the existence of the gap. Perhaps reformulate this as an example

File: design/design.md [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+systems including proof-of-stake blockchains where the amount of honest stake
+significantly exceeds the amount of malicious stake.
+
+Additionally, ALBA supports a decentralized setting where knowledge of the data
```
Comment (rrtoledo) [inline design/design.md:39]: You can be more direct and make it easier to understand:
"where the dataset is decentralized and a single party aggregates the elements to generate a proof".

File: design/design.md [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+elements is spread out across multiple provers that need to communicate to
+create a proof.
+More concretely, the provers send a message to an aggregator who in turn
+will produce a proof for the verifier.
```
Comment (rrtoledo) [inline design/design.md:43]: It is confusing that the aggregator generate a proof and not the provers. I strongly suggest to change the terminology.

File: design/design.md [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+of the library.
+We aim to create a high quality reference implementation of ALBA sufficient
+for accurately prototyping larger protocols utilizing this primitive.
+Applications that could benefit, at least in the Cardano universe, include
+Mithril, Peras, Leios and cross-chain bridges.
+These use cases will generally guide the development of our library, though
+our aim is to also allow the open-source community to study and experiment
+with ALBA.
+
+### Overview of ALBA
```
Comment (rrtoledo) [inline design/design.md:18]: This is way too long, we already have a README and doc that you can link.
You should write the overall concept quickly and refer to them.

File: design/design.md [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+through a lottery as used, for example, by Algorand's sortition or the Mithril
+protocol (todo: add link).
+The lottery has slightly better communication complexity but a significantly
+worse proof size.
+
+See the [website](https://alba.cardano-scaling.org) or
+the [paper](https://eprint.iacr.org/2023/1655) to get a more detailed overview
+of ALBA.
+
+### Applications of ALBA
```
Comment (rrtoledo) [inline design/design.md:65]: Again, way too long, and too Cardano centered. This is a design document and not application. 
You should remove most of this. You can mention them and describe them succinctly, also refer to paper. But that's it.
You should also provide different examples, or overall applications as we would like this library to be used by other people than Cardano.

File: design/design.md [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+partner chains like Midnight.
+
+Applications of ALBA are not limited to Cardano and blockchains in general.
+For instance, the [paper](https://eprint.iacr.org/2023/1655) describes how
+this primitive can be used to optimize the prover running time of
+a Universally Composable SNARK.
+
+## Requirements
+
+We intend to make an implementation of ALBA that can be used in
```
Comment (rrtoledo) [inline design/design.md:122]: Remove all Cardano applications. Just say we want the implementation to be generic enough to be used by any project.

File: design/design.md [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+Mithril, Peras, Leios, cross-chain bridges and potentially other protocols
+inside and outside the Cardano universe.
+Our implementation will mostly be guided by our main "clients":
+Mirthril, Peras and Leios.
+
+ALBA is a generic tool that lets one prove knowledge of any kind of data
```
Comment (rrtoledo) [inline design/design.md:128]: it's knowledge of a _bound_, be precise.

File: design/design.md [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+elements, but
+as noted in (todo: alba applications section), these three applications
```
Comment (rrtoledo) [inline design/design.md:130]: This is not important here, we should already state that the library shall be application agnostic.

File: design/design.md [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+don't need the full expressivity of ALBA, they really need
+a multisignature scheme, where the data elements would specifically be
+signatures.
+Thus there is question of whether we want our library to provide a general
```
Comment (rrtoledo) [inline design/design.md:134]: No, that is wrong.

We will provide an _example_ that will be Alba used as a multisignature scheme.
This will not be a separate library but a separate crate / folder in this repo.

File: design/design.md [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+as many applications as possible.
+This means that there is only a generic notion of "data elements" and no notion
+of signatures, VRF, etc.
+2. We will use the generic ALBA library to implement a multisignature scheme,
+either as a small example in this repository or a full-fledged library in a
+separate repository.
+Either way, the multi-signature implementation would serve as a flagship
+example of how the ALBA library can be used.
+
+The requirements for our (general) ALBA implementation can be roughly split
```
Comment (rrtoledo) [inline design/design.md:147]: These requirements are very, very loose. I would rather have goals written: 
- the library must implement correctly the protocols as written in the paper as a reference implementation.
- the code must be easy to follow
...

Comment (rrtoledo) [inline design/design.md:147]: Did you discuss with annyone before writing these requirements?

File: design/design.md [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+in four groups: 
+1. definitions: what do the ALBA schemes we implement guarantee, under what
+assumptions?
+2. correctness: does our implementation provide the same guarantees of
+soundness and completeness as the ALBA schemes on paper?
+3. performance: proof size, communication complexity, computational complexity.
+4. interfaces: the applications should be able to use our
+implementation relatively easily
+
+### Definitions
```
Comment (rrtoledo) [inline design/design.md:157]: I do not understand what you mean by "definitions". API? Interface? Be more precise and clear

File: design/design.md [hunk 21]
```diff
@@ -0,0 +1,414 @@
...
+
+The definitions requirement can be stated as follows.
+Taking an ALBA protocol we implement, can you construct (on paper) a larger
+protocol, such as Mithril, and mathematically prove its guarantees such as
+completeness and soundness?
+Currently, ALBA schemes satisfying the definitions in the ALBA paper cannot
+be used to construct Mithril, Peras and Leios.
+
+First, all three require a multisignature scheme secure against
```
Comment (rrtoledo) [inline design/design.md:166]: Remove the following, whole paragraphs. We want to cover these cases, not focus on it. You can say much more succinctly what you want without using them as points.

File: design/design.md [hunk 22]
```diff
@@ -0,0 +1,414 @@
...
+For Mithril, if the data elements are signatures, the weight oracle would
+have to report for any signature the amount of stake of the creator of the
+signature.
+Clearly, this is impossible.
+Even if the data element was a pair of public key and signature, the weight
+oracle would have no way of verifying that the public key owns any stake.
+Again, research is needed to construct the right ALBA scheme on
+paper before implementing it.
+
+### Correctness
```
Comment (rrtoledo) [inline design/design.md:191]: There will _always_ be a discrepancy because of types and what's not. There is no "ideal", we will implement the scheme with that goal using best practices, you may note that there is a discrepancy between theory and practice.
You may list where the errors are, but be more concise, and don't use "tiny".

File: design/design.md [hunk 23]
```diff
@@ -0,0 +1,414 @@
...
+The following is a list of known discrepancies that prevent us from making
+the rigorous guarantees about soundness and completeness errors.
+1. tiny rounding errors in parameter derivation due to the use of
+floating-point numbers
+2. tiny errors when using a uniformly random hash output to sample distributions
+needed for ALBA such as Bernoulli, uniform, binomial, Poisson
+3. lack of a rigorous analysis of the $H_1$ optimization described in
+section 3.3 of the paper
+
+### Performance
```
Comment (rrtoledo) [inline design/design.md:213]: Stop talking about Peras, Mithril and Alba.
This wil be talks withing IOHK with the relevant teams. It should not be in here.

File: design/design.md [hunk 24]
```diff
@@ -0,0 +1,414 @@
...
+In that case, further techniques can be considered such as SNARKs and
+taking the stake distribution into account (todo: see fait-accompli).
+
+These protocols' requirements about communication and computational complexity
+are not currently known.
+However, the communication complexity of the implemented ALBA schemes will not greatly
+exceed that of the simple lottery, and the prover running time is expected to be
+in the order of a second.
+
+### Interfaces
```
Comment (rrtoledo) [inline design/design.md:246]: Stop talking about Peras and Leios. You can just write that we will implement bindings

**Review by rrtoledo — CHANGES_REQUESTED** [review]
Halfway review (up to Proposed design choices):

This is supposed to be a design document, not a README or documentation.
I think you spent too much time writing about the overview of ALBA or applications (which is already presented in these) and selling ALBA merits. You can dive much earlier into the requirements and design choices which is the goal of the document.

This is a design document for a reference implementation of ALBA. You can list (a couple of times), Cardano projects that may use it but more is not necessary, even counter productive. The document should be short, and appeal to a wide audience. You should remove most of the discussions involbing Peras, Mithril, Leios, etc..

The document is not precise and is too long. You do not say that ALBA is a lower bound at the beginning. Your requirements are very loose. How do we know if we achieved them? And their naming is not explicit. Do another pass and be concise. Many paragraphs can be rewritten in a sentence or two

File: design/design.md [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+(Thus, the simple lottery in the decentralized weighted setting is equivalent
+to Algorand's sortition.)
+Finally, we can create centralized weighted ALBA schemes by remodeling the
+decentralized weighted schemes as centralized;
+this is currently the only known efficient way to implement weights.
+
+We will implement Telescope Bounded,
+as well as, the simple lottery in all four settings (centralized+decentralized,
+unweighted+weighted).
+The following is the full list of eight implemented schemes.
```
[the remark this answers is not in the document]
Comment (rrtoledo) [reply design/design.md:290]: I would just write that something like "we implement the bounded and lottery schemes to cover the following settings"

File: design/design.md [hunk 26]
```diff
@@ -0,0 +1,414 @@
...
+* Centralized unweighted Telescope Bounded (section 3.2.2)
+* Decentralized unweighted Telescope Bounded (section 3.2.2 + 4.2)
+* Centralized+decentralized weighted Telescope Bounded (section 3.2.2 + 4.2 + 5)
+* Centralized+decentralized unweighted simple lottery (section 4.1)
+* Centralized+decentralized weighted simple lottery (section 4.1 + 5)
+
+#### Discussion
```
Comment (rrtoledo) [inline design/design.md:297]: You could be more concise in the whole section. Many paragraphs can be rewritten with fewer lines and words. You do not needto explain why something is more efficient when it is obvious for instance.

I also find the style a bit incoherent since you first list the reasons "in english" (the following reasons. ...), then you have long sentence to list all for lottery, and finally you have an itemize for the redundancy.

File: design/design.md [hunk 27]
```diff
@@ -0,0 +1,414 @@
...
+2. because they might be more efficient now or when optimizations are
+implemented in the future; for instance, unweighted variants can be made
+more efficient due to no lottery and less complexity, and in the decentralized
+setting, an unweighted version could be more efficient due to not needing to
+sample the binomial distribution.
+
+We should be able to reuse most of the logic between different schemes anyway,
+so the overhead shouldn't be large.
+
+### Choice of hash function
```
Comment (rrtoledo) [inline design/design.md:329]: Likewise, you can be much more concise. 

I would not write at all about domain separation. We can say we shall implement it for Alba, and that users using Alba should also implement it on the application level. But that's pretty much it.

This is a design document for Alba, we should not explain too much about the rest, the part on using different hash functions and correlation is in particular: this is not in the scope of the document, and you actually may need to use different ones:a hash function for a hashmap has different requirements than one for zerocoin.

File: design/design.md [hunk 28]
```diff
@@ -0,0 +1,414 @@
...
+### Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+### Parameter calculation
+ALBA has two kinds of parameters: user-friendly parameters such as
+$\lambda_{sec}$, $\lambda_{rel}$, $n_p$, $n_f$,
+and algorithm specific parameters such as $u$, $d$ and $q$.
```
Comment (rrtoledo) [inline design/design.md:377]: Add perhaps "internal" as we say this in different places (comments, docs)

File: design/design.md [hunk 29]
```diff
@@ -0,0 +1,414 @@
...
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+### Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+### Parameter calculation
```
Comment (rrtoledo) [inline design/design.md:374]: @djetchev  and I agreed that versioning would be suffficient for evolving the protocol so I would remove the last paragarph about backward compatibility.

As for recommendation about the derivation and hardcore, this is out of the scope and not necessarily good. We could also recommend to cache the parameters (if we use alba many times with changing parameters) for instance.

File: design/design.md [hunk 30]
```diff
@@ -0,0 +1,414 @@
...
+The ALBA functions such as `prove` and `verify` will thus receive as input
+the algorithm specific parameters.
+
+The rationale is that we want to be able to evolve parameter derivation
+(e.g., to optimize parameters for performance or make the derivation logic
+more rigorous) while maintaining backward compatibility with previous versions
+of the library.
+
+## Roadmap
+We have already implemented the Telescope Bounded scheme
```
Comment (rrtoledo) [inline design/design.md:394]: You can rewrite this in one sentence, "we will implement 1, 2 and 3 in this order". In parallel, we will implement benchmarks and so on...

File: design/design.md (hunk 4 above)
Comment (rrtoledo) [inline design/design.md:393]: You can be more concise.

The communication with clients must disappear altogether. This is outside of the scope of this document, and is a different (non github) task to integrate Alba in Cardano's subprotocols. Likewise for the example and mathematical analysis, we should only say that the library will be maintained with research (not that we do it) and potentially enriched with more examples

**Review by rrtoledo — CHANGES_REQUESTED** [review]
I think we have different view on the document. I see this document as a very concise, straightforward and precise document focusing on the design of the repository: what is the goal of the library (that's missing), what we shall implement, design choices, perhaps example choice. To name a few things:
- We can mention some applications, but we should not delve too deep into them or mention them in other places than the application section. We can perfectly argue for generics by saying that we want the lib to be as general as possible and application agnostic.
- We shall not talk about how Alba will be used. This is the work of an architect, cryptographer... We can give examples and suggestions of course, but we should not go further.
We already have documentation and comments all over the code so we should make the most of them by referring them here instead of rewriting Alba overview. We do not want to bore the reader, just give them the information they want. To us, this document should be a clear guidelines, and reminder, of our goals.

You should furthermore try to emphasize more the core of the document (what we implement, roadmap), especially if you are reluctant to remove text or rewrite the sections. You can separate the texts, use more lists (as @curiecrypt mentionned)...

**Comment by curiecrypt** [comment]
In addition to the points I mentioned in my review, I would like to add the following:
- The protocol explanation sections could be more concise, referencing the relevant parts of the paper for additional details (as you've done in some parts).
- The emphasis on the projects Peras, Leios, and Mithril should be reduced. Focusing too heavily on these projects makes it seem like ALBA is a protocol exclusively designed for and applicable to them. Instead, we should present ALBA in a more general context, highlighting its broader potential and flexibility. This approach would allow readers to better understand the protocol's capabilities and explore its boundaries.
- The document's purpose should be articulated more clearly. For the introduction section, the `Overview of ALBA` could be more concise, and `Applications of ALBA` could adopt a broader perspective. Most importantly, the introductory paragraph should be more explanatory. It would be helpful to describe the document's structure, outlining what is covered in each section, the rationale for this organization, and the intended flow of information.
- The roadmap should be more detailed and specific. We can collaborate to determine its exact content.
- Success criteria should be included and connected to the roadmap. These criteria should be quantitative and provide clear benchmarks for evaluating the project's progress and outcomes.

File: design/design.md (hunk 29 above)
Comment (tolikzinovyev) [reply design/design.md]: Deleted this section.

**Review by tolikzinovyev — COMMENTED** [review]
Thank you everyone for the review! I addressed most of the comments.

I see your point about focusing too much on Peras, Leios, etc. I talk about them because it's much easier to have some applications in mind rather than developing something very abstract. This is why in the first paragraph I said that these applications "will guide the development of our library". Do we agree with this statement? Off top of my head, I propose rewording the document to make it more clear that this library is for general use.

@curiecrypt what would you like to see for the success criteria?

File: design/design.md (hunk 9 above)
Comment (rrtoledo) [reply design/design.md]: I have always had an issue with this word. We usually say tight or loose bounds, Alba was the first time I hear a bound to be approximate hence I am not sure what the meaning would be.

File: design/design.md (hunk 12 above)
Comment (rrtoledo) [reply design/design.md:42]: Cann you say "parties" perhaps?  It is general enough and also used in the paper :grimacing:

File: design/design.md (hunk 16 above)
Comment (rrtoledo) [reply design/design.md:136]: Wouldn't this be _how_ your primitive is designed to solve the prof of lower bound rather than the goal of the primitive though?

File: design/design.md (hunk 20 above)
Comment (rrtoledo) [reply design/design.md:168]: I am not fond of the title (as it is not clear) and not sure about what you say. Is it that you want the design of the implementation to be generic enough?

File: design/design.md (hunk 22 above)
Comment (rrtoledo) [reply design/design.md:206]: Because of precision. I could ask for a security paramater equals to E and no computer would be able to give me the exact result. 

Regardless, our goal is to balance correctness, readability and security of the protocol. This is a reference implementation hence we care about the protocol being correct, not the rounding and other errors.

tiny: I have not seen it in papers, I usually see "not significant", so I have been assuming it to be rather informal

File: design/design.md (hunk 27 above)
Comment (rrtoledo) [reply design/design.md:355]: I am not saying that domain separation is not important but that we should focus on our repository: we implemented domain separation as security good practice. We can recommend to also use it in the wider application, but it is not the goal of the document to say what should, or shouldn't be done, there. Hence I would keep it short.

File: design/design.md (hunk 19 above)
Comment (rrtoledo) [reply design/design.md:155]: So how do we know if the requirements are satisfied or not?

File: design/design.md (hunk 13 above)
Comment (rrtoledo) [reply design/design.md:19]: This document is about the design of our implementation and the choices we have made. Someone reading this should already know what Alba is (or seen the README on the first page).

File: design/design.md (hunk 5 above)
Comment (tolikzinovyev) [reply design/design.md:12]: I mentioned voting protocols and multisignatures.

File: design/design.md (hunk 27 above)
Comment (tolikzinovyev) [reply design/design.md:356]: But this section directly affects the repository / library implementation: whether we allow the user to provide a custom hash function and what the interface should be. While the exact interface is not specified here, this sections sets guidance for the right interface.

File: design/design.md (hunk 14 above)
Comment (tolikzinovyev) [reply design/design.md:66]: I reworded the beginning the section to emphasize that alba is general purpose. Additionally, the end of the section again says that it's not only for cardano. I think it's useful to have concrete applications in mind to have direction.

File: design/design.md (hunk 20 above)
Comment (rrtoledo) [reply design/design.md:163]: So you are saying that the repository should be an accurate implementation of the paper? You could rename this paragraph then "Consistency (to the paper/specifications)"

File: design/design.md [hunk 31]
```diff
@@ -0,0 +1,414 @@
...
+Cross-chain bridges are protocols allowing one blockchain to verify
+information on another blockchain.
+In essense this is the same problem that the Mithril solves with a potentially
+added constraint that on-chain verification must be computationally cheap.
+Whereas Mithril is specific to the Cardano blockchain, one can envision
+a similar protocol, powered by ALBA, for other cryptocurrencies including
+partner chains like Midnight.
+
+Applications of ALBA are not limited to Cardano and blockchains in general.
+For instance, the [paper](https://eprint.iacr.org/2023/1655) describes how
```
[the remark this answers is not in the document]
Comment (curiecrypt) [reply design/design.md:118]: It still gives me the same feeling, since we start the sentence with "For instance". 
However, it is not a big issue. The reader will eventually see which paper it is when they click on the link.

File: design/design.md [hunk 32]
```diff
@@ -0,0 +1,414 @@
...
+
+In our implementation, we will let users provide their own function.
+Hash functions evolve, old ones get broken (SHA-1), and so our ALBA
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
```
[the remark this answers is not in the document]
Comment (curiecrypt) [reply design/design.md:371]: You could clarify by stating that the security and cryptographic properties of user-defined hash functions are also the responsibility of the user.

**Review by curiecrypt — CHANGES_REQUESTED** [review]
This document should include more bullet lists rather than long paragraphs. However, the content has matured and better satisfies its goal. 
I gave a couple of suggestions but addressing them is totally up to you. 
Thanks for the effort.

File: design/design.md (hunk 4 above)
Comment (tolikzinovyev) [reply design/design.md]: Thanks, I used some of this! As for success criteria, I'm not if we want to be so formal and strict. For example, "interface usability scores". Are we actually going to ask people to give a score 1-10. Or quantitative criteria for benchmarks, are we going to define performance numbers that we need to satisfy?

File: design/design.md [hunk 33]
```diff
@@ -0,0 +1,419 @@
...
+We aim to create a high quality reference implementation of ALBA sufficient
+for accurately prototyping larger protocols utilizing this primitive such as
+protocols for voting and multisignatures schemes.
+In the Cardano universe, projects that can benefit from ALBA include
+Mithril, Peras, Leios and cross-chain bridges.
+While these use cases will guide the development of our library, one can
+imagine other applications / protocols that can utilize ALBA, and
+our aim is to allow the open-source community to study and experiment
+with it.
+
```
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply design/design.md:19]: done!

**Review by rrtoledo — APPROVED** [review]
LGTM

**Review by curiecrypt — APPROVED** [review]
LGTM 🚀
